rs/src/connections/host_relay.rs

```diff
@@ -0,0 +1,1085 @@
+            .await;
+
+        match tunnel_port {
+            // created the port:
+            Ok(_) => {}
+            // the port's already registered, nothing we need to do:
+            Err(HttpError::ResponseError(e)) if e.status_code == 409 => {}
+            Err(e) => {
+                return Err(TunnelError::HttpError {
+                    error: e,
+                    reason: "failed to add port to tunnel",
+                })
+            }
+        }
+
+        let (tx, rx) = mpsc::unbounded_channel();
+        self.ports_tx.send_modify(|v| {
+            v.insert(n, tx);
+        });
+
+        Ok(rx)
+    }
+
+    /// Adds a new port to the tunnel and forwards TCP/IP connections made
+    /// over that port to the local machine. Calling this method multiple times
+    /// with the same port will result in an error.
+    pub async fn add_port(&self, port_to_add: &TunnelPort) -> Result<(), TunnelError> {
+        let rx = self.add_port_raw(port_to_add).await?;
+
+        tokio::spawn(forward_port_to_tcp(
+            SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), port_to_add.port_number).into(),
+            rx,
+        ));
+
+        Ok(())
+    }
+
+    /// Removes a port from the tunnel connection. Any channel returned from
+    /// `add_port_raw`, and any connections made within `add_port`, will close
+    /// shortly after this is called.
+    pub async fn remove_port(&self, port_number: u16) -> Result<(), TunnelError> {
+        self.mgmt
+            .delete_tunnel_port(&self.locator, port_number, NO_REQUEST_OPTIONS)
+            .await
+            .map_err(|e| TunnelError::HttpError {
+                error: e,
+                reason: "failed to remove port from tunnel",
+            })?;
+
+        self.ports_tx.send_modify(|v| {
+            v.remove(&(port_number as u32));
+        });
+
+        Ok(())
+    }
+
+    fn make_ssh_server(keypair: russh_keys::key::KeyPair) -> Server {
+        let c = russh::server::Config {
+            connection_timeout: None,
+            auth_rejection_time: std::time::Duration::from_secs(5),
+            keys: vec![keypair],
+            window_size: 1024 * 1024 * 64,
+            preferred: russh::Preferred::COMPRESSED,
+            limits: russh::Limits {
+                rekey_read_limit: usize::MAX,
+                rekey_time_limit: Duration::MAX,
+                rekey_write_limit: usize::MAX,
+            },
+            ..Default::default()
+        };
+
+        let config = Arc::new(c);
+        Server { config }
+    }
+
+    async fn make_ssh_client(
+        rw: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    ) -> Result<
+        (
+            russh::client::Handle<Client>,
+            mpsc::UnboundedReceiver<ChannelOp>,
+        ),
+        russh::Error,
+    > {
+        let config = russh::client::Config {
+            anonymous: true,
+            window_size: 1024 * 1024 * 64,
```

Comment:
The window size for the outer SSH session (the SSH client session between host and relay) should be larger than the window size for the inner SSH sessions (the SSH server sessions between clients and hosts), to prevent a single client from filliing the host's window and thereby blocking other clients. In https://github.com/microsoft/dev-tunnels/pull/135, I used 5 MB for the outer session and 1 MB (default) for the inner sessions; the relay services is also updated to use 5 MB for the host session.

Eventually this protocol will change to avoid layering SSH, and then the window size will be less important.